Color Module - First Draft (#147)
design-tokens/community-group: technical-reports/color/color-type.md

# Color type

## Format

> Colors can be represented through various formats. For color tokens, the `$type` property MUST be set to the string `color`. The `$value` property can then be used to specify the details of the color, including color space, alpha settings, and more. The `$value` object contains the following properties:

Reviewer: Note: resolving this comment (and others) that are old. Going to make new comments to reflect recent color group meetings and internal feedback shared among DTCG editors. Resolving right now means “feedback heard and accepted; ALSO there will be imminent changes to this PR.

> - `$hex` (required): the hex color to use as the default or a guaranteed fallback

Reviewer: Suggestion: `$hex`, `$colorSpace`, `$name` etc should probably drop the `$` prefix, right? For all other tokens, `$value` objects contain normal non-prefixed keys. Which feels more expected because outside the token, `$value` is a special reserved word that marks any object as a token. But inside the `$value`, a strict schema is enforced where arbitrary properties are no longer allowed (so no benefit to prefixing keys).

Apologies if there’s a discussion I missed or if this suggestion isn’t helpful.

- `$colorSpace` (optional): An object detailing an alternative color space to be used to interpret the color, if supported

> The `$colorSpace` object has the following properties:

Reviewer: Question: should we explain why it’s an array, and why the order matters (following CSS Color 4 spec)? Or is that too much information?

> - `$name` (required): the name of the color space (either rgb, srgb, hsl, or lch)

Reviewer 2: I agree, since `colorSpace` is optional, it can have a more distinct nomenclature instead of `name`

> - `$components` (required): the non-alpha pieces of the color, listed as an array of floating-point numbers or integers

Reviewer: Oh yes this **definitely needs clarification!** Because no, values won’t always be between `0-1` if we follow [CSS Color Module 4](https://www.w3.org/TR/css-color-4/). There are 2 reasons:

1. **Degrees**. As in the case of HSL, the range is 0°–360°. This could probably be normalized to `1` but I think most would agree that fails the goal of this spec being human writable since that’s nonstandard
2. **Device-independent colorspaces.** In the case of colorspaces like LAB/LCH and OKLAB/OKLCH, they describe many colors that are outside the range of any device existing today, and so there’s no “maximum.”

For the latter, CSS Color Module 4 still does define allowable ranges which are still far beyond what most devices can support; I believe they just put reasonable caps on values that decades of hardware improvements still wouldn’t catch up to. Here are some of the values:

- Lab: L [0 … 100], a [-125 … 125], b [-125 … 125]
- Lch: L [0 … 100], c [0 … 150], h [0 … 360]
- Oklab: L [0 … 1], a [-0.4 … 0.4], b [-0.4 … 0.4]
- Oklab: L [0 … 1], c [0 … 0.4], h [0 … 360]
- XYZ: 🤔 for some reason this is normalized [0 … 1] but not sure why

The general approach of “match CSS Color Module 4 as much as possible” has been discussed in meetings but I think it would be good to explicate this as much as possible. As far as I know, the color proposal wouldn’t deviate from Color Module 4 in any way. cc @resource11

> - `$components` (required): the non-alpha pieces of the color, listed as an array of floating-point numbers or integers

Reviewer: Also `alpha` in all cases is always normalized [0 … 1], and is optional (default `1` if omitted).

> - `$components` (required): the non-alpha pieces of the color, listed as an array of floating-point numbers or integers

Reviewer 2: @drwpow yeah, if no color space is omitted, and no other color function is used, opacity/alpha channel (rgba) should always default to normalized 0-1

- `$alpha` (optional): the alpha component of the color, listed as a floating-point number integer in the range of 0 to 1. If omitted, color is assumed to be fully opaque

## Hex - required support

> For the color value, the required fallback format to represent color through design tokens is a hex or hex8 value. A hex triplet is a 6-digit, 24 bit, hexadecimal number that represents Red, Green, and Blue values as `#RRGGBB`. An eight-character hex will include the alpha value in the last 2 characters. If no alpha value is specified, it is assumed the color if fully opaque.

Reviewer: Agreed. Hex values have started to be used for Display P3, but the intent is limited to sRGB.

> For the color value, the required fallback format to represent color through design tokens is a hex or hex8 value. A hex triplet is a 6-digit, 24 bit, hexadecimal number that represents Red, Green, and Blue values as `#RRGGBB`. An eight-character hex will include the alpha value in the last 2 characters. If no alpha value is specified, it is assumed the color if fully opaque.

Reviewer 3: ```suggestion
For the color value, the required fallback format to represent color through design tokens is a hex or hex8 value. A hex triplet is a 6-digit, 24 bit, hexadecimal number that represents Red, Green, and Blue values as `#RRGGBB`. An eight-character hex will include the alpha value in the last 2 characters. If no alpha value is specified, it is assumed the color if fully opaque. The sRGB colorspace MUST be used for interpreting fallback hex values.
```

          >)
        </li>
      </ul>
    </td>
  </tr>
</table>

<aside class="example">

```json
{
  "Majestic magenta": {
    "value": "#ff00ff80",
    "type": "color"
  },
  "Simple sage": {
    "value": "#abcabc80",
    "type": "color"
  }
}
```

</aside>

Then, the output variables may look like:

<aside class="example">

```scss
// colors-hex.scss
$majestic-magenta: #ff00ff80;
$simple-sage: #abcabc80;

// colors-rgba.scss
$majestic-magenta: rgba(255, 0, 255, 0.5);
$simple-sage: rgba(171, 202, 188, 0.5);
```

</aside>

### LCH (Lightness Chroma Hue)

Formatted in L (lightness), C (chroma), H (hue) and an optional (A) alpha. Hue values range from 0 to 360, saturation and lightness are percentage values that go from 0% to 100%, and alpha value can range from 0 and 1 (i.e 0.5) or a percentage (i.e 50%) where 1 or %100 is full opacity (which is the default value if a value isn’t provided).

<table>
  <tr>
    <td><strong>Pros</strong></td>
    <td><strong>Cons</strong></td>
  </tr>
  <tr>
    <td>
      <ul>
        <li>Access to 50% more colors (P3 color space)</li>
        <li>Colors more perceptually uniform</li>
      </ul>
    </td>
    <td>
      <ul>

>         <li>No fully supported (only safari)</li>

Reviewer 4: ```suggestion
        <li>Not fully supported (only safari)</li>
```

Reviewer 4: Just a nitpick. It would be nice if the Con sections mentioning the lack of browser support were consistent.

Something like 

> "Limited Browser Support (Safari)"

Consistent verbiage and browsers with support are listed.

# Color Type

Colors can be represented through various formats. For color, the type property must be set to the string “color”. For the value, the most common format to represent color through design tokens is a hex triplet. A hex triplet is a 6-digit, 24 bit, hexadecimal number that represents Red, Green, and Blue values as #RRGGBB.   

> 

Reviewer 5: THis is a comment

<table>
  <tr>
   <td><strong>Pros</strong>
   </td>
   <td><strong>Cons</strong>
   </td>
  </tr>
  <tr>
   <td>Easily recognized among tools and browsers
<p>
??
   </td>
   <td>Cannot specify alpha value for opacity
   </td>
  </tr>
</table>

For example, initially color tokens may be defined as such:

```
{
  "Majestic magenta": {
    "value": "#ff00ff",
    "type": "color"
  },
  "Simple sage": {
    "value": "#abcabc",
    "type": "color"
  }

> }

Author: Update examples

    "$value": "#ff00ff",
    "$type": "color"
  },
  "Simple sage": {
    "$value": "#abcabc",
    "$type": "color"
  }
}
```

</aside>

Then, the output from a tool’s conversion to HSL may look something like:

<aside class="example">

```scss
// colors-hex.scss
$majestic-magenta: #ff00ff;
$simple-sage: #abcabc;

// colors-hsl.scss
$majestic-magenta: ​hsl(300, 100%, 50%);
$translucent-shadow: ​hsl(153, 23%, 73%);
```

</aside>

## Other value options

### RGBA

> Formatted in R (red), G (green), B (blue) and (A) alpha. Red, green, and blue values can range from 0 to 255 and alpha values can range from 0 and 1 (i.e 0.5) or a percentage (i.e 50%) where 1 or %100 is full opacity.

Reviewer 5: ```suggestion
Formatted in R (red), G (green), B (blue) and (A) alpha. Red, green, and blue values can range from `0` to `255`, and the alpha value ranges from `0` to `1` (such as `0.5`) or a percentage (such as `50%`) where `1` or `100%` is full opacity.
```

For example, initially color tokens may be defined as such:

<aside class="example">

```json
{
  "Majestic magenta": {
    "$value": {
      "red": 255,
      "green": 0,
      "blue": 255,
      "alpha": 1
    },
    "$type": "color"
  },
  "Simple sage": {
    "$value": {
      "red": 171,
      "green": 202,
      "blue": 188,
      "alpha": "50%"
    },
    "$type": "color"
  }
}
```

</aside>

Then, the output from a tool’s conversion to RGBA may look something like:

<aside class="example">

```scss
// colors-rgba.scss
$majestic-magenta: rgba(255, 0, 255, 1);
$translucent-shadow: rgba(171, 202, 188, 50%);
```

</aside>

### HSL

Formatted in H (hue), S (saturation), L (lightness) and an optional (A) alpha. Hue values range from 0 to 360, saturation and lightness are percentage values that go from 0% to 100%, and alpha value can range from 0 and 1 (i.e 0.5) or a percentage (i.e 50%) where 1 or 100% is full opacity (which is the default value if a value isn’t provided).

| Pros                                                  | Cons                                  |
| ----------------------------------------------------- | ------------------------------------- |

> | It is easy to understand and compare to other formats | Not supported in all browsers (IE 11) |

Reviewer 5: >Not supported in all browsers (IE 11)

Source?

<aside class="example">

```json
{
  "Majestic magenta": {
    "h": 300,
    "s": "100%",
    "l": "50%",
    "a": "100%",
    "$type": "color"
  },
  "Simple sage": {
    "h": 100,
    "s": "27%",
    "l": "57%",
    "a": "100%",
    "$type": "color"
  }
}
```

</aside>

Then, the output variables may look like:

<aside class="example">

```scss
// colors-rgba.scss
$majestic-magenta: hsl(300, 100%, 50%, 1);
$simple-sage: hsl(100, 27%, 57%);
```

</aside>

### Hex8

> Hex8 uses two extra digits, known as the alpha value, to change the transparency of the color. The format follows `#RRGGBBAA`. [Learn more about alpha values in hex.](https://www.digitalocean.com/community/tutorials/css-hex-code-colors-alpha-values#adding-an-alpha-value-to-css-hex-codes)

Reviewer 5: ```suggestion
Hex8 uses two extra digits, known as the alpha value, to change the transparency of the color. The format follows `#RRGGBBAA`. [Learn more about alpha values in hex codes](https://www.digitalocean.com/community/tutorials/css-hex-code-colors-alpha-values#adding-an-alpha-value-to-css-hex-codes).
```

| --------------------------------- | ------------------------------------------------------------------------------------------------------------ |
| Can define alpha value with color | Less commonly used                                                                                           |
|                                   | Alpha value is not immediately obvious (needs calculation)                                                   |
|                                   | Not available in older versions of Internet Explorer ([caniuse reference](https://caniuse.com/css-rrggbbaa)) |

<aside class="example">

```json
{
  "Majestic magenta": {
    "$value": "#ff00ff80",
    "$type": "color"
  },
  "Simple sage": {
    "$value": "#abcabc80",
    "$type": "color"
  }
}
```

</aside>

Then, the output variables may look like:

<aside class="example">

```scss
// colors-hex.scss
$majestic-magenta: #ff00ff80;
$simple-sage: #abcabc80;

// colors-rgba.scss
$majestic-magenta: rgba(255, 0, 255, 0.5);
$simple-sage: rgba(171, 202, 188, 0.5);
```

</aside>

### LCH (Lightness Chroma Hue)

> Formatted in L (lightness), C (chroma), H (hue) and an optional (A) alpha. Hue values range from 0 to 360, saturation and lightness are percentage values that go from 0% to 100%, and alpha value can range from 0 and 1 (i.e., 0.5) or a percentage (i.e., 50%) where 1 or %100 is full opacity (which is the default value if a value isn’t provided).

Reviewer 5: ```suggestion
Formatted in L (lightness), C (chroma), H (hue) and an optional (A) alpha. Hue ranges from `0` to `360`, saturation and lightness are percentage values that range from `0%` to `100%`, and the optional alpha value ranges from `0` and `1` (such as `0.5`) or as a percentage (such as `50%`) where `1` or `100%` is full opacity (which is the default value if a value isn’t provided).
```

| Pros                                       | Cons                                                                        |
| ------------------------------------------ | --------------------------------------------------------------------------- |
| Access to 50% more colors (P3 color space) | Colors more perceptually uniform, so it can be harder to distinguish values |

> |                                            | Limited browser support (Safari only)                                       |

Reviewer 5: >Limited browser support (Safari only)

Now supported in Chrome, Edge, Safari, and Firefox behind a flag.
